### ISSUE TITLE
Problems adding devices using: stable_release-7-92-0

### ISSUE THREAD

**Issue by joshStillerman** [MDSplus/mdsplus#1911]
Brian -

  I did a clean install of that stable on UBUNTU18 see:

https://github.com/MDSplus/mdsplus/issues/1910

and was able to add an acq132 device from MitDevices:

Looking at my ENVIRONMENT variables I see:

> jas@elk:~$ printenv | grep PATH
> LD_LIBRARY_PATH=/usr/local/mdsplus/lib:/etc/alternatives/jre/lib/amd64:/usr/local/cmod/lib64:/usr/local/cmod/lib32
> MDS_PYDEVICE_PATH=/usr/local/mdsplus/pydevices
> MDSMSG_PATH=/usr/local/mdsplus/lib
> PATH=/home/jas/CodeSourcery/Sourcery_CodeBench_Lite_for_Xilinx_GNU_Linux/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/mdsplus/bin:/usr/local/cmod/bin:/home/jas/bin
Looking in the MDS_PYDEVICE_PATH for files I see:

> jas@elk:/usr/local/mdsplus$ ls /usr/local/mdsplus/pydevices/
> MitDevices
> jas@elk:/usr/local/mdsplus$ ls /usr/local/mdsplus/pydevices/MitDevices/
> a3248.py      ACQ196_FTP.glade    acq216_ftp.py    acq.py          DT132.glade   dt196b.pyc          DT216B.glade    dtdo32.py     picam.pyc         transport.py
> a3248.pyc     ACQ196.glade    acq216_ftp.pyc    acq.pyc       dt132.py        dt196.py          dt216b.py    dtdo32.pyc    raspicam.py     transport.pyc
> ACQ132.glade  acq196.py        ACQ216.glade    CP7452.glade  dt132.pyc     dt196.pyc          dt216b.pyc    idl.py          raspicam.pyc     _version.py
> acq132.py     acq196.pyc    acq216.py    cp7452.py     dt196a.py     Dt200WriteMaster.py   dt216.py    idl.pyc       raspi_rgb.py     _version.pyc
> acq132.pyc    acq200.py        acq216.pyc    cp7452.pyc    dt196a.pyc    Dt200WriteMaster.pyc  dt216.pyc    __init__.py   raspi_rgb.pyc
> acq196ao.py   acq200.pyc    acq_ftp.py    dt100.py      DT196B.glade  dt216a.py          dtao32.py    __init__.pyc  timeSignature.py
> acq196ao.pyc  ACQ216_FTP.glade    acq_ftp.pyc    dt100.pyc     dt196b.py     dt216a.pyc          dtao32.pyc    picam.py      timeSignature.pyc
I notice that I still have __init__.py and .pyc  I will remove them and see if things still work:  - ANSWER YES
> jas@elk:/usr/local/mdsplus/pydevices/MitDevices$ sudo mv __init__.py* ~/
> [sudo] password for jas:
> jas@elk:/usr/local/mdsplus/pydevices/MitDevices$ cd
> jas@elk:~$ mdstcl
> TCL> edit test
> TCL> add node acq3 /model=acq132
> TCL> 

And I can add the Joerger device you tried also:

> jas@elk:~$ mdstcl
> TCL> edit test
> TCL> add node test/model=joerger_tr812
> TCL> 
Without a MDS_DEVICES environment variable the tdi fun MitDevices()  returns the devices in our shared image:

> jas@elk:~$ tdic
> TDI> MitDevices()
> [["DC1394A      ","MitDevices   "], ["DT196AO      ","MitDevices   "], ["RETICON120   ","MitDevices   "], ["A14          ","MitDevices   "], ["A3204        ","MitDevices   "], ["B5910A       ","MitDevices   "], ["HV1440       ","MitDevices   "], ["HV1443       ","MitDevices   "], ["HV4032       ","MitDevices   "], ["HV4032A1     ","MitDevices   "], ["INCAA16      ","MitDevices   "], ["INCAA4       ","MitDevices   "], ["INCAA6       ","MitDevices   "], ["INCAA_TR10   ","MitDevices   "], ["J1819        ","MitDevices   "], ["J221         ","MitDevices   "], ["JOERGER_TR16 ","MitDevices   "], ["JOERGER_TR612","MitDevices   "], ["JOERGER_TR812","MitDevices   "], ["JRG_TR1612   ","MitDevices   "], ["JRG_ADC32A   ","MitDevices   "], ["L6810        ","MitDevices   "], ["L6810A       ","MitDevices   "], ["L6810B       ","MitDevices   "], ["L6810C       ","MitDevices   "], ["L8501        ","MitDevices   "], ["L8590        ","MitDevices   "], ["L8590_MEM    ","MitDevices   "], ["L8590_SCLR   ","MitDevices   "], ["L8818        ","MitDevices   "], ["L8828        ","MitDevices   "], ["MIT_ENCODER  ","MitDevices   "], ["MPB__DECODER ","MitDevices   "], ["T2812        ","MitDevices   "], ["T2814        ","MitDevices   "], ["T2824        ","MitDevices   "], ["T2825        ","MitDevices   "], ["T2860        ","MitDevices   "], ["T4012        ","MitDevices   "], ["CHS_A14      ","MitDevices   "], ["DC1394       ","MitDevices   "], ["DSP2904      ","MitDevices   "], ["DT200        ","MitDevices   "], ["DT200DA      ","MitDevices   "], ["DT_ACQ16     ","MitDevices   "], ["MATROX       ","MitDevices   "], ["PARAGON_HIST ","MitDevices   "], ["PARAGON_RPT  ","MitDevices   "], ["PREAMP       ","MitDevices   "], ["UMCCD        ","MitDevices   "], ["U_OF_M_SPECT ","MitDevices   "], ["A12          ","MitDevices   "], ["B2408        ","MitDevices   "], ["B3224        ","MitDevices   "], ["EC727        ","MitDevices   "], ["FERA         ","MitDevices   "], ["H908         ","MitDevices   "], ["H911         ","MitDevices   "], ["H912         ","MitDevices   "], ["HM650        ","MitDevices   "], ["J412         ","MitDevices   "], ["JOERGER_ADC  ","MitDevices   "], ["JOERGER_ADCP ","MitDevices   "], ["JOERGER_CG   ","MitDevices   "], ["JOERGER_DAC16","MitDevices   "], ["L2232        ","MitDevices   "], ["L2256        ","MitDevices   "], ["L2415        ","MitDevices   "], ["L3512        ","MitDevices   "], ["L4202        ","MitDevices   "], ["L8100        ","MitDevices   "], ["L8201        ","MitDevices   "], ["L8206        ","MitDevices   "], ["L8210        ","MitDevices   "], ["L8212_04     ","MitDevices   "], ["L8212_08     ","MitDevices   "], ["L8212_16     ","MitDevices   "], ["L8212_32     ","MitDevices   "], ["MIT_CLOCK    ","MitDevices   "], ["MIT_DCLOCK   ","MitDevices   "], ["MIT_DECODER  ","MitDevices   "], ["MIT_GATE     ","MitDevices   "], ["MIT_GCLOCK   ","MitDevices   "], ["MIT_PULSE    ","MitDevices   "], ["MIT__CLOCK   ","MitDevices   "], ["MIT__DCLOCK  ","MitDevices   "], ["MIT__GATE    ","MitDevices   "]]
> TDI> 
Even with an MDS_DEVICES environment variable, I can not get it to show my python devices.... I have to look into this....

So...

If you change MDS_PYDEVICE_PATH to be /usr/local/mdsplus/pydevices  can you add our one of our python and or C devices ?

I am putting this in as an issue on github - which n the future is a better way to raise these things.

-Josh

On 1/28/20 7:21 PM, Brian Nelson wrote:
> Hi Josh et al.,
>
> After upgrading to Ubuntu 18.04 and installing MDSplus stable version 7.92.0, we are apparently unable to add any new Devices to a Tree.
>
> Demonstration:  I removed all our UWDevices (which gave the "Device xxx already imported" errors I mentioned previously) from our env vars
>
> nelson@fuze:~$ env | grep MDS | grep DEV
> MDS_PYDEVICE_PATH=/usr/local/UWmdsplus/tdi/UWDevices/Python;/usr/local/mdsplus/pydevices
> MDS_DEVICES=UWDevices:MitDevicesSub:KbsiDevices:RfxDevices:W7xDevices
>
> nelson@fuze:~$ export MDS_PYDEVICE_PATH=/usr/local/mdsplus/pydevices
> nelson@fuze:~$ export MDS_DEVICES=MitDevices:KbsiDevices:RfxDevices:W7xDevices
>
> Then I verified MitDevices.fun works in TDI (truncated some lines shown by "..."):
>
> nelson@fuze:~$ tdic
> TDI> MitDevices()
> [["DC1394A      ","MitDevices   "],
> ["DT196AO      ","MitDevices   "], 
> ["RETICON120   ","MitDevices   "], 
> ["A14          ","MitDevices   "], 
> ["A3204        ","MitDevices   "], 
> ["B5910A       ","MitDevices   "], 
> ["HV1440       ","MitDevices   "], 
> ...
> ["JOERGER_TR812","MitDevices   "], 
> ["JRG_TR1612   ","MitDevices   "], 
> ["JRG_ADC32A   ","MitDevices   "], 
> ...
> ["MIT__DCLOCK  ","MitDevices   "], ["MIT__GATE    ","MitDevices   "]]
>
> and tried to add an MitDevice JOERGER_TR812 to a new Tree and mdstcl hangs:
>
> nelson@fuze:~$ mdstcl
> TCL> edit/new nelson/shot=-1
> TCL> add node test/model=joerger_tr812
>
> Any ideas? 
>
> Thanks,
> Brian
>
> On Tue, Jan 21, 2020 at 12:06 PM Stillerman,Joshua A. <jas@psfc.mit.edu> wrote:
>
>     Brian -
>
>       I think that if you device records contain something like:
>
>     Build_Conglom("__python__", "ACQ435ST", *, "from acq435st import ACQ435ST")
>
>     and the environment variable PY_DEVICE_PATH includes the directory with your code in it (';' separated I believe) it should all just work.  You no longer need (or want) __init__.py
>
>     -Josh
>     On 1/21/20 2:44 PM, Brian Nelson wrote:
>>     Hi All,
>>
>>     After finally upgrading Ubuntu from 14.04 -> 18.04, and upgrading MSDplus, we found that MDSplus Python Devices had changed.
>>
>>     For our "UWdevices", I changed the __init__.py from the "from XXXX import XXXX" to the (updated) "_mimport("XXXX", "XXXX")" style, which allowed Python Devices in existing Trees to load and work properly.
>>
>>     However, we cannot add new Devices to a Tree, it spits out several "already imported" notices, then hangs (see below).  This occurs for *any* attempt to add a Device (model) Node, whether it is in UWDevices or some other Devices group.
>>
>>     ****
>>     nelson@fuze:~$ mdstcl
>>     TCL> edit nelson/new/shot=-1
>>     TCL> add node acq480_080/model=acq480_v2
>>     Device 'L4222_PRO' from 'L4222_PRO' already imported from 'L4222_PRO'
>>     Device 'J221_GPIB' from 'J221_GPIB' already imported from 'J221_GPIB'
>>     Device 'ACQMC' from 'ACQMC' already imported from 'ACQMC'
>>     Device 'L4222_GPIB' from 'L4222_GPIB' already imported from 'L4222_GPIB'
>>     Device 'ACQMC_V2' from 'ACQMC_V2' already imported from 'ACQMC_V2'
>>     Device 'J221A_GPIB' from 'J221A_GPIB' already imported from 'J221A_GPIB'
>>     Device 'TEKSCOPE_8CH' from 'TEKSCOPE_8CH' already imported from 'TEKSCOPE_8CH'
>>     Device 'ACQ480' from 'ACQ480' already imported from 'ACQ480'
>>     ^CKeyboardInterrupt
>>     ****
>>
>>     Is this something people have experienced in the change to the new Devices "_mimport()" use?  Is there documentation of how the new(er) system works?
>>
>>     I'm pretty sure I do *not* have any explicit Python "import" calls for any of our UWDevices.
>>
>>     Thanks,
>>     Brian
>>
>>     -- 
>>     Brian A. Nelson
>>     Affiliate Professor
>>     Retired Research Professor
>>     Department of Electrical & Computer Engineering
>>     University of Washington, Seattle
>>
>>     _______________________________________________
>>     Mdsplus mailing list
>>     Mdsplus@lists.psfc.mit.edu
>>     To subscribe, send a message to Mdsplus-subscribe@lists.psfc.mit.edu
>>     To unsubscribe, send a message to Mdsplus-unsubscribe@lists.psfc.mit.edu
>>     https://lists.psfc.mit.edu/mailman/listinfo/mdsplus
>     -- 
>     Joshua Stillerman
>     Research Engineer
>     MIT Plasma Science and Fusion Center
>     617.253.8176
>     jas@mit.edu
>     _______________________________________________
>     Mdsplus mailing list
>     Mdsplus@lists.psfc.mit.edu
>     To subscribe, send a message to Mdsplus-subscribe@lists.psfc.mit.edu
>     To unsubscribe, send a message to Mdsplus-unsubscribe@lists.psfc.mit.edu
>     https://lists.psfc.mit.edu/mailman/listinfo/mdsplus
>
>
>
> -- 
> Brian A. Nelson
> Affiliate Professor
> Retired Research Professor
> Department of Electrical & Computer Engineering
> University of Washington, Seattle
-- 
Joshua Stillerman
Research Engineer
MIT Plasma Science and Fusion Center
617.253.8176
jas@mit.edu

**Comment by nelson-zapenergy** [comment]
Hi Josh,

"It's not you, it's me."  

I had defined our MDS_PATH to look at our "local" TDI files first; then look in /usr/local/mdsplus/tdi.

When I changed MDS_PATH to reverse the search order to first look at "stock" TDI and then our "local" TDI, *and* defining MDS_DEVICES to include our UWDevices, I can add the MitDevices acq132, the joerger_tr812, and our own devices.

I will check what in our "local" TDI is messing things up.

Thanks for your help.
